File: js/euler87.js
function euler87(){
    let limit = 5e7;
    const sieved = new Array(Math.floor(Math.sqrt(limit))).fill(false);
    const squares = [], cubes = [], fourths = [];
    for (let i = 2; i < sieved.length; i++){
        if (!sieved[i]){
            squares.push(i ** 2);
            if (i ** 3 < limit) cubes.push(i ** 3);
            if (i ** 4 < limit) fourths.push(i ** 4);
            for (let j = 2 * i; j < sieved.length; j += i){
                sieved[j] = true;
            }
        }
    }
    const sums = new Set();
    for (let fourth of fourths){
        for (let cube of cubes){
            if (fourth + cube >= limit) break;
            for (let square of squares){
                let total = fourth + cube + square;
                if (total >= limit) break;
                sums.add(total);
            }
        }
    }
    return sums.size;
}

console.log(euler87());

File: go/euler87.go
package main
import (
    "fmt"
    "math"
)

func sieve(n int) []int {
    sieved := make([]bool, n + 1)
    primes := []int{}
    for i := 2; i <= n; i++ {
        if !sieved[i] {
            primes = append(primes, i)
            for j := 2 * i; j <= n; j += i {
                sieved[j] = true
            }
        }
    }
    return primes
}

func main() {
    n := 50000000
    maxSqrt := int(math.Sqrt(float64(n)))
    maxCbrt := int(math.Cbrt(float64(n)))
    maxFtrt := int(math.Sqrt(math.Sqrt(float64(n))))
    
    squares := []int{}
    cubes := []int{}
    fourths := []int{}
    
    for _, p := range sieve(maxSqrt) {
        sq := p * p
        if p <= maxFtrt {
            fourths = append(fourths, sq * sq)
        }
        if p <= maxCbrt {
            cubes = append(cubes, p * sq)
        }
        squares = append(squares, sq)
    }
    
    triples := make(map[int]bool)
    for _, fourth := range fourths {
        for _, cube := range cubes {
            if fourth + cube >= n {
                break
            }
            for _, square := range squares {
                triple := fourth + cube + square
                if triple >= n {
                    break
                }
                triples[triple] = true
            }
        }
    }
    
    fmt.Println(len(triples))
}
